### DISCUSSION TITLE
Help needed with Custom auth.

### DISCUSSION THREAD

**Discussion by Benjythebee** [supabase/supabase#15612] [Questions]
Heya, I need help setting up custom auth with RLS;

I used the [following tutorial](https://medium.com/@gracew/using-supabase-rls-with-a-custom-auth-provider-b31564172d5d) to get me started but I quickly realised it was outdated.

I'm stuck at generating a JWT and starting a session for a user given that JWT. Since the removal of `auth.setAuth()` I'm not too sure how to achieve my end goal in nextJS.

I have a *public.users* table with an `id` column which is a foreign key duplicate of the `id` column of *auth.users*.
RLS is enabled and with the following UPDATE check: `(auth.uid() = id)`.

Here is my code:

**/api/auth/login.ts** (simplified for readability)
```typescript
const handler = async (req: NextApiRequest, res: NextApiResponse) => {
const { email } = req.body
	const supabaseServerClient = createPagesServerClient({
		req,
		res,
	});

        let user = await getUserByEmail(email, supabaseServerClient);
       if(user){ // User already exists at this point
		const newSecret = generateSecret();  // generats a random string
		const token = genCookie(user,req,res) // See function genCookie below (generates the JWT)

		supabaseServerClient.functions.setAuth(token)
		supabaseServerClient.realtime.setAuth(token)

                const {data:users,error} = await supabase
	         .from("users") // public.users table
	           .update({secret: newSecret })
	             .eq("email", email)
	            .select();
		if(!user){
			return res.status(422).json({ error: "Could not update user " + reconstructedAddress});
		}
           return res.send(users[0])
     }else{
      // error
     }
}

```

```js
const genCookie = (user:SupabaseUser,req:IncomingMessage,res:ServerResponse)=>{
	const token = jwt.sign({ id:user.id, email:user.email }, process.env.JWT_SECRET ); //JWT_SECRET is from the supabase dashboard

	return token
}
````

My issue is that the policy check clearly doesn't work since the query below returns error = `null` and users= `[]` (yet the user exists). So I assume my understanding of auth is completely wrong.
```
               const {data:users,error} = await supabase
	         .from("users") // public.users table
	           .update({secret: newSecret })
	             .eq("email", email)
	            .select();
```

I'm really confused at how I can make auth work with custom auth... I'd love some guidance.
Also worth noting, I am using NextJS pages directory (not App directory);

```
    "@supabase/auth-helpers-nextjs": "^0.7.2",
    "@supabase/auth-helpers-react": "^0.4.0",
    "@supabase/supabase-js": "^2.24.0",
```

**Comment by j4w8n** [comment]
When creating your jwt, replace the `id` key with `sub`. This is what `auth.uid()` checks in the policy.

**Reply by Benjythebee** [reply]
I've changed it to 
```
jwt.sign({ sub: user.id, email:user.email }, process.env.JWT_SECRET ); //JWT_SECRET is from the supabase dashboard
```
 and it made no difference saddly...

**Reply by Benjythebee** [reply]
and the jwt token generated seems valid because if I run 
```
 const userFromToken = await supabaseServerClient.auth.getUser(token)
 console.log(userFromToken)
```
it shows the user; now sure what to do about auth

**Comment by Benjythebee** [comment]
Ok, I've found a partial solution, it looks like the module needs to be updated so that the `rest` headers are also updated;

I changed my code to add a one-line hack:
```js
		supabaseServerClient.functions.setAuth(token)
		supabaseServerClient.realtime.setAuth(token)
		;(supabaseServerClient as any).rest.headers.Authorization = `Bearer ${token}`
```
and now I can successfully edit the user's data;

However I'm still not sure how to obtain a session from there

**Reply by Benjythebee** [reply]
@kiwicopple sorry for the name drop, seems like a PR will be needed for the `auth-helpers` repo; one that adds a `setAuth()` function similar to `functions.setAuth()` but for the `rest` headers

**Reply by j4w8n** [reply]
You can set the global header for the supabase client, which should cover you here.

**Reply by Benjythebee** [reply]
The auth-helpers module do not expose a global supabase client I think. As far as I know what you're saying means I'd have to re-declare
```js
	const supabaseServerClient = createPagesServerClient({
		req,
		res,
	},	{	options:{
			global:{
				headers:{'Authorization':`Bearer ${jwt}`}
			}
		}}
		);
```
every single time the jwt changes; 

So in this example for the `login.ts` code, you're suggesting I override the supabaseServerClient variable with a new client?

**Reply by j4w8n** [reply]
No I was just saying to set the global headers option, as your code has. You were saying that you had to manually set the `rest` header, and this should solve that.
